pkg/acl/acl.go

```diff
@@ -0,0 +1,408 @@
+func (a *ACL) SetFileACLDefault(path string) error {
+	acl_set_file, err := getSymbolPointer(a.lib.handle, "acl_set_file")
+	if err != nil {
+		return err
+	}
+
+	cpath := C.CString(path)
+	defer C.free(unsafe.Pointer(cpath))
+
+	ret, errno := C.my_acl_set_file(acl_set_file, cpath, C.ACL_TYPE_DEFAULT, a.a)
+	if ret < 0 {
+		return fmt.Errorf("error calling acl_set_file: %v", errno)
+	}
+
+	return nil
+}
+
+// Valid checks whether the ACL is valid.
+func (a *ACL) Valid() (bool, error) {
+	acl_valid, err := getSymbolPointer(a.lib.handle, "acl_valid")
+	if err != nil {
+		return false, err
+	}
+
+	ret := C.my_acl_valid(acl_valid, a.a)
+	if ret < 0 {
+		return false, nil
+	}
+	return true, nil
+}
+
+// AddBaseEntries adds the base ACL entries from the file permissions.
+func (a *ACL) AddBaseEntries(path string) error {
+	fi, err := os.Lstat(path)
+	if err != nil {
+		return err
+	}
+	mode := fi.Mode().Perm()
+	var r, w, x bool
+
+	// set USER_OBJ entry
+	r = mode&userRead == userRead
+	w = mode&userWrite == userWrite
+	x = mode&userExec == userExec
+	if err := a.addBaseEntryFromMode(TagUserObj, r, w, x); err != nil {
+		return err
+	}
+
+	// set GROUP_OBJ entry
+	r = mode&groupRead == groupRead
+	w = mode&groupWrite == groupWrite
+	x = mode&groupExec == groupExec
+	if err := a.addBaseEntryFromMode(TagGroupObj, r, w, x); err != nil {
+		return err
+	}
+
+	// set OTHER entry
+	r = mode&otherRead == otherRead
+	w = mode&otherWrite == otherWrite
+	x = mode&otherExec == otherExec
+	if err := a.addBaseEntryFromMode(TagOther, r, w, x); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *ACL) createEntry() (*Entry, error) {
+	acl_create_entry, err := getSymbolPointer(a.lib.handle, "acl_create_entry")
+	if err != nil {
+		return nil, err
+	}
+
+	var e C.acl_entry_t
+
+	rv, _ := C.my_acl_create_entry(acl_create_entry, &a.a, &e)
+	if rv < 0 {
+		return nil, fmt.Errorf("unable to create entry")
+	}
+	return &Entry{a, e}, nil
+}
+
+func (a *ACL) addBaseEntryFromMode(tag Tag, read, write, execute bool) error {
+	e, err := a.createEntry()
+	if err != nil {
+		return err
+	}
+	if err = e.setTag(tag); err != nil {
+		return err
+	}
+	p, err := e.getPermset(a)
+	if err != nil {
+		return err
+	}
+	if err := p.addPermsFromMode(read, write, execute); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (entry *Entry) getPermset(a *ACL) (*Permset, error) {
+	acl_get_permset, err := getSymbolPointer(a.lib.handle, "acl_get_permset")
+	if err != nil {
+		return nil, err
+	}
+
+	var ps C.acl_permset_t
+	rv, _ := C.my_acl_get_permset(acl_get_permset, entry.e, &ps)
+	if rv < 0 {
+		return nil, fmt.Errorf("unable to get permset")
+	}
+	return &Permset{a, ps}, nil
+}
+
+func (entry *Entry) setTag(t Tag) error {
+	acl_set_tag_type, err := getSymbolPointer(entry.a.lib.handle, "acl_set_tag_type")
+	if err != nil {
+		return err
+	}
+
+	rv, _ := C.my_acl_set_tag_type(acl_set_tag_type, entry.e, C.acl_tag_t(t))
+	if rv < 0 {
+		return fmt.Errorf("unable to set tag")
```

Comment:
Same old. Probably a separate function for this might be handy, like this totally untested code:
```
func wrapErrno(msg string, errno uintptr) error {
	return errwrap.Wrap(errors.New(msg), syscall.Errno(errno))
}
```